event-dispatcher/src/main/java/io/sk8s/event/dispatcher/KafkaConsumerMonitor.java

```diff
@@ -0,0 +1,193 @@
+import java.util.Properties;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.springframework.util.StopWatch;
+
+/**
+ * Tracks current and end offset (and hence lag) for a set of topic/consumerGroup pairs.
+ *
+ * @author Eric Bottard
+ */
+public class KafkaConsumerMonitor implements Closeable {
+
+	private final String host;
+
+	private final Map<String, Consumer<?, ?>> consumersByGroup = new HashMap<>();
+
+	private final String monitoringConsumerGroupID = "monitoring_consumer_" + UUID.randomUUID().toString();
+
+	private volatile Consumer<?,?> logEndOffsetTrackingConsumer;
+
+	private final Set<TopicGroup> tracked = new HashSet<>();
+
+	private final static Collection<TopicPartition> FOR_ALL_PARTITIONS = Collections.emptyList();
+
+	public KafkaConsumerMonitor(String host) {
+		this.host = host;
+	}
+
+	@Override
+	public synchronized void close() {
+		for (TopicGroup topicGroup : tracked) {
+			stopTracking(topicGroup.group, topicGroup.topic);
+		}
+		if (logEndOffsetTrackingConsumer != null) {
+			logEndOffsetTrackingConsumer.close();
+		}
+	}
+
+	public synchronized void beginTracking(String group, String topic) {
+		if (logEndOffsetTrackingConsumer == null) {
+			logEndOffsetTrackingConsumer = createNewConsumer(monitoringConsumerGroupID);
+		}
+		consumersByGroup.computeIfAbsent(group, k -> createNewConsumer(group));
+		tracked.add(new TopicGroup(topic, group));
+		reassignPartitions();
+	}
+
+	public synchronized void stopTracking(String group, String topic) {
+		tracked.remove(new TopicGroup(topic, group));
+		if (tracked.stream().noneMatch(tg -> tg.group.equals(group))) {
+			consumersByGroup.remove(group).close();
+		}
+		reassignPartitions();
+	}
+
+	private void reassignPartitions() {
+		logEndOffsetTrackingConsumer.assign(
+				tracked.stream()
+						.flatMap(tg -> logEndOffsetTrackingConsumer.partitionsFor(tg.topic).stream()
+								.map(pi -> new TopicPartition(tg.topic, pi.partition()))
+						).collect(Collectors.toSet()));
+	}
+
+	public Map<TopicGroup, List<Offsets>> compute() {
+		logEndOffsetTrackingConsumer.seekToEnd(FOR_ALL_PARTITIONS);
+
+		return tracked.stream()
+				.collect(Collectors.toMap(
+						Function.identity(),
+						tg -> {
+							List<PartitionInfo> partitionInfos = logEndOffsetTrackingConsumer.partitionsFor(tg.topic);
+							return partitionInfos.stream()
+									.map(pi -> new TopicPartition(tg.topic, pi.partition()))
+									.map(
+											tp -> {
+												OffsetAndMetadata committed = consumersByGroup.get(tg.group).committed(tp);
+												if (committed == null) {
+													committed = new OffsetAndMetadata(0L);
+												}
+												long position = logEndOffsetTrackingConsumer.position(tp);
+												return new Offsets(tp.partition(), position,
+														committed.offset());
+											})
+									.collect(Collectors.toList());
+						}));
+	}
+
+	static class TopicGroup {
```

Comment:
this name is confusing since it sounds like *a group of Topics*... maybe this is just the `ConsumerGroup`? (or ConsumerKey)

Reply:
It's really a dumb Tuple used as a data structure in the set of what is tracked.
If functions were supporting multiple inputs (which they could, as we discussed) this would evolve to 1 Group+N topics

Reply:
yeah, my point is that it's not representing a group of Topics yet that's what it suggests

maybe `Subscription`?

Reply:
...or `Bindings`? ;)